### PR TITLE
document recursive solver

### PR DESCRIPTION
**Pull request by nikomatsakis**
OK, I don't know how to do much better than this!

You do want to view it with `mdbook serve` to get a feel for what it says.

cc @flodiebold, @zaharidichev

### CODE REVIEW

File: book/src/recursive.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+impl B for u32 { }
+
+impl A for i32 { }
+```
+
+In the recursive solver, with a goal of `Implemented(Vec<?X>: A)`, we
+recursively try to prove `Implemented(?X: A)` and get ambiguity, and we get
+stuck there.
+
+The SLG solver in contrast starts by exploring `?X = u32` and finds
```
Comment (shepmaster) [inline book/src/recursive.md:93]: Link this to the slg page?

File: book/src/recursive/inductive_cycles.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+to the same goal, S and S1. This implies two possibilities:
+
+* If S == S1, then in fact there is a unique solution, so we can return `Provable(S)`.
+* If S != S1, then we know there are two solutions, which means that there is
+  not one unique solution, and hence the correct result is **ambiguous**,
+  and in fact we can just stop and return right now.
+
+This technique is very similar to the traditional Prolog technique of handling
+cycles, which is called **tabling**. The difference between our solution and
+table is that we are always looking for a unique solution, whereas Prolog (like
```
Comment (shepmaster) [inline book/src/recursive/inductive_cycles.md:84]: Table -> tabling?

File: book/src/recursive/inductive_cycles.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+the [SLG solver]) tries to enumerate all solutions (i.e., in prolog, solving a
```
Comment (shepmaster) [inline book/src/recursive/inductive_cycles.md:85]: Prolog

File: book/src/recursive/search_graph.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+[cycles]: ./inductive_cycles.md
+[coinduction]: ./coinduction.md
+[stack]: ./stack.md
+[`SearchGraph`]: http://rust-lang.github.io/chalk/chalk_solve/recursive/search_graph/struct.SearchGraph.html
+[`DepthFirstNumber`]: http://rust-lang.github.io/chalk/chalk_solve/recursive/search_graph/struct.DepthFirstNumber.html
+[`Node`]: http://rust-lang.github.io/chalk/chalk_solve/recursive/search_graph/struct.Node.html
+[`stack_depth`]: http://rust-lang.github.io/chalk/chalk_solve/recursive/search_graph/struct.Node.html#structfield.stack_depth
+
+The role of the [`SearchGraph`] is to store information each goal that we are
```
Comment (shepmaster) [inline book/src/recursive/search_graph.md:15]: “About”?

File: book/src/recursive/stack.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+The stack itself never stores the goal associated with a particular entry. That
+information is found in the [search graph], which will be covered in detail in
+the next section. For now it suffices to say that the search graph maps from
+"some goal that we are currently solving" to "information about that goal", and
+one of the bits of information is the [`StackDepth`] of its entry on the stack
+(if any).
+
+Therefore, when we are about to start solving some (canonical) goal G, we can
+detect a cycle by checking in the searh graph to see whether G has an associated
```
Comment (shepmaster) [inline book/src/recursive/stack.md:71]: Search

**Comment by nikomatsakis** [comment]
Thanks @shepmaster -- fixed!

**Pull request merged**